### PR TITLE
Middleware monitoring config

### PR DESCRIPTION
**Pull request by aidenkeating**
Rebase of #385 on #384 

@sedroche 

## Verification Steps for 3scale

1. Run this installation branch.
2. Go to the Prometheus UI. (`prometheus-route`/config  in the `middleware-monitoring` namespace) and make sure a new config is present.

config

3. Ensure there are 2 new targets at `prometheus-route`/targets
targets
4. Ensure you can run a query v the new targets. e.g. `nginx_http_connections{_3scale_component_element="staging"}` at `prometheus-route`/graph
graph

## Verification Steps for federation

1. Run the installer and ensure it completes (should already be done in 3scale)
2. Ensure Kube State Metrics is not in the `middleware-monitoring` namespace
3. Ensure an `openshift-monitoring-federation` target exists in prometheus and is up
screen shot 2019-02-21 at 10 27 17
4. Ensure Alerts are passing (they may be pending for a while)
screen shot 2019-02-08 at 21 15 24
5. Run the uninstaller and ensure it completes

### CODE REVIEW

File: evals/roles/middleware_monitoring_config/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+---
+- include: ./kube_state_metrics_alerts.yml
+- include: ./prometheus_additional_scrape_config.yml
```
Comment (sedroche) [inline evals/roles/middleware_monitoring_config/tasks/main.yml]: are we missing a `- include: ./openshift_monitoring_federation.yml`?

**Comment by sedroche** [comment]
Added my verification steps from https://github.com/integr8ly/installation/pull/384 in this description

File: evals/roles/middleware_monitoring_config/tasks/main.yml (hunk 1 above)
Comment (aidenkeating) [reply evals/roles/middleware_monitoring_config/tasks/main.yml]: I've merged the two files, apart from retrieving some secret data they did the same thing

File: evals/roles/middleware_monitoring_config/tasks/prometheus_additional_scrape_config.yml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    shell: oc get prometheus application-monitoring -n {{ middleware_monitoring_namespace }}
+    register: monitoring_prometheus_check
+    until: monitoring_prometheus_check.rc == 0
+    delay: 5
+    retries: 50
+
+  - name: Patch the middleware monitoring prometheus with additional scrape config
+    shell: "oc patch prometheus {{ prometheus_cr_name }} -n {{ middleware_monitoring_namespace }} --patch='{{ additional_scrape_config_patch | to_json }}' --type='merge'"
+
+  # - name: Delete the temp yaml files
```
Comment (sedroche) [inline evals/roles/middleware_monitoring_config/tasks/prometheus_additional_scrape_config.yml]: Do we not need this?

File: evals/roles/middleware_monitoring_config/tasks/main.yml (hunk 1 above)
Comment (sedroche) [reply evals/roles/middleware_monitoring_config/tasks/main.yml]: Gotcha :+1:

File: evals/roles/middleware_monitoring_config/tasks/prometheus_additional_scrape_config.yml (hunk 2 above)
Comment (aidenkeating) [reply evals/roles/middleware_monitoring_config/tasks/prometheus_additional_scrape_config.yml]: Removed it while debugging an error after the rebase, I've uncommented it now

**Comment by aidenkeating** [comment]
@sedroche I've included expanded verification steps from #384

File: evals/roles/middleware_monitoring_config/tasks/prometheus_additional_scrape_config.yml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+  - name: Check if the middleware monitoring prometheus CR is available
+    shell: oc get prometheus application-monitoring -n {{ middleware_monitoring_namespace }}
+    register: monitoring_prometheus_check
+    until: monitoring_prometheus_check.rc == 0
+    delay: 5
+    retries: 50
+
+  - name: Patch the middleware monitoring prometheus with additional scrape config
+    shell: "oc patch prometheus {{ prometheus_cr_name }} -n {{ middleware_monitoring_namespace }} --patch='{{ additional_scrape_config_patch | to_json }}' --type='merge'"
```
Comment (sedroche) [inline evals/roles/middleware_monitoring_config/tasks/prometheus_additional_scrape_config.yml]: This is my code (and I thought it was in there) but...
There should be a check here from previous patches I have seen.
```python
  - name: Patch the middleware monitoring prometheus with additional scrape config
    shell: "oc patch prometheus {{ prometheus_cr_name }} -n {{ middleware_monitoring_namespace }} --patch='{{ additional_scrape_config_patch | to_json }}' --type='merge'"
    register: prometheus_cr_patch
    failed_when: prometheus_cr_patch.stderr != '' and 'not patched' not in prometheus_cr_patch.stderr
    changed_when: prometheus_cr_patch.rc == 0
```
This allows reruns of the script

**Comment by aidenkeating** [comment]
@sedroche I see the 3scale config being added to the prometheus config secret

```
- job_name: 3scale-apicast-pods
  kubernetes_sd_configs:
  - role: pod
    namespaces:
      names:
      - 3scale
  relabel_configs:
  - action: labelmap
    regex: __meta_kubernetes_pod_label_(.+)
    replacement: _$1
  - source_labels:
    - __meta_kubernetes_namespace
    action: replace
    target_label: kubernetes_namespace
  - source_labels:
    - __meta_kubernetes_pod_name
    action: replace
    target_label: kubernetes_pod_name
  - source_labels:
    - __meta_kubernetes_pod_annotation_prometheus_io_scrape
    action: keep
    regex: true
  - source_labels:
    - __address__
    - __meta_kubernetes_pod_annotation_prometheus_io_port
    action: replace
    regex: ([^:]+)(?::\d+)?;(\d+)
    replacement: $1:$2
    target_label: __address__
```
* I see the 3scale target showing up in Prometheus
screen shot 2019-02-21 at 10 42 08
* I can query the 3scale data (using `nginx_http_connections{_3scale_component_element="staging"}`)
screen shot 2019-02-21 at 10 43 23

**Comment by sedroche** [comment]
Code looks good

**Comment by sedroche** [comment]
Verifying

**Pull request merged**